client/electron/go_vpn_tunnel.ts

```diff
@@ -310,17 +319,39 @@ class GoTun2socks {
   }
 
   /**
-   * Checks connectivity and exits with the string of stdout.
+   * Checks connectivity of the server specified in `config`.
+   * Checks whether proxy server is reachable, whether the network and proxy support UDP forwarding
+   * and validates the proxy credentials.
    *
+   * @returns A boolean indicating whether UDP forwarding is supported.
    * @throws ProcessTerminatedExitCodeError if tun2socks failed to run successfully.
    */
-  checkConnectivity() {
+  async checkConnectivity(config: TransportConfigJson): Promise<boolean> {
     console.debug('[tun2socks] - checking connectivity ...');
-    return this.process.launch([
+    const output = await this.process.launch([
       '-transport',
-      JSON.stringify(this.transportConfig),
+      JSON.stringify(config),
       '-checkConnectivity',
     ]);
+    // Only parse the first line, because sometimes Windows Crypto API adds warnings to stdout.
+    const outObj = JSON.parse(output.split('\n')[0]);
+    if (outObj.tcp) {
+      throw new Error(outObj.tcp);
+    }
+    if (outObj.udp) {
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Fetches a dynamic key from the given URL.
+   * @param url The URL to be fetched.
+   * @throws ProcessTerminatedExitCodeError if we failed to fetch the config.
+   */
+  fetchConfig(url: string): Promise<string> {
```

Comment:
Same, move this to a standalone function. We should add stuff to this `GoTun2socks` class if not needed. (really it should be deleted).

Reply:
Done